### PR TITLE
LSP: Add Go To Definition support for `Ast::ModuleAccess`

### PR DESCRIPTION
**Pull request by jansul**
Fairly straightforward one, this adds Go To Definition support for `Ast::ModuleAccess` as demonstrated in the video below:

https://github.com/mint-lang/mint/assets/1927518/b76a52e9-acae-40d9-80a2-4549b9e20d90

While allowed by the language, some of these examples will error when running them [in the browser](https://sandbox.mint-lang.com/sandboxes/lN4-HdUp6_JfXw). If you like I can have a look at why next if you like?

As part of this PR, I have removed the file `src/parsers/constant_variable.cr` and replaced it with the new method for parsing constants introduced in https://github.com/mint-lang/mint/pull/594. This old version still had the bug which allowed invalid constants.

### CODE REVIEW

**Review by Sija — APPROVED** [review]
That's rad! Love the test coverage ❤️

**Comment by jansul** [comment]
OK, turns out it was less straightforward than I thought :sweat_smile: 

I forgot that you can:
 * Have multiple modules with the same name
 * Modules can share their names with enums/records

This PR should now work correctly with both. For the former, it will now link to all instances of a module. Here's how that looks like in VSCode:

https://github.com/mint-lang/mint/assets/1927518/a9db14c8-ec42-45bc-bcf0-1f28c4861cac

File: src/ls/definition.cr [hunk 1]
```diff
@@ -4,7 +4,7 @@ module Mint
     class Definition < LSP::RequestMessage
       property params : LSP::TextDocumentPositionParams
 
-      def execute(server) : LSP::LocationLink | LSP::Location?
+      def execute(server) : Array(LSP::LocationLink | LSP::Location) | LSP::LocationLink | LSP::Location | Nil
```
Comment (Sija) [inline src/ls/definition.cr:7]: I'd prefer a more unified return type `Array(LSP::LocationLink | LSP::Location)`

**Review by Sija — CHANGES_REQUESTED** [review]

File: src/ls/definition.cr (hunk 1 above)
Comment (jansul) [reply src/ls/definition.cr:7]: I thought the above matched the [specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_typeDefinition), but it turned out it doesn’t!

The return type for “Go to definition” can be any of `Location | Location[] | LocationLink[] | null`, but a singular `LocationLink` is not allowed - so we are currently out of spec 😅

Simplifying it to `Array(LSP::LocationLink | LSP::Location)` might be a good option, will have to have a think about how best to refactor all the definition methods.

Comment (gdotdesign) [reply src/ls/definition.cr:7]: > Simplifying it to `Array(LSP::LocationLink | LSP::Location)` might be a good option, will have to have a think about how best to refactor all the definition methods.

IMO, that can be in a separate PR.

**Review by gdotdesign — APPROVED** [review]

**Review by Sija — APPROVED** [review]

**Pull request merged**